### PR TITLE
Pre/post run commands added

### PR DESCRIPTION
**Pull request by andre2007**
After the run of an executable there are artefacts created like docs.json (-Xdocs.json). This artefacts should be further processed. postGenerateCommands and postBuildCommands runs before the exectutable is executed.

postRunCommands helps in this scenario. preRunCommands is added for symmetry.

### CODE REVIEW

**Comment by andre2007** [comment]
Docs update https://github.com/dlang/dub-docs/pull/10

**Comment by andre2007** [comment]
Ping

File: test/issue1396-pre-post-run-commands/dub.sdl [hunk 1]
```diff
@@ -0,0 +1,3 @@
+name "test"
+preRunCommands "echo \"pre-run\n\" >> test.txt"
```
Comment (wilzbach) [inline test/issue1396-pre-post-run-commands/dub.sdl]: Not sure why you need the newline

File: test/issue1396-pre-post-run-commands.sh [hunk 2]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
+
+. $(dirname "${BASH_SOURCE[0]}")/common.sh
+cd ${CURR_DIR}/issue1396-pre-post-run-commands
+rm -rf .dub
```
Comment (wilzbach) [inline test/issue1396-pre-post-run-commands.sh]: Isn't this dangerous when the testsuite is running in parallel (though that's not really supported as of now)

File: test/issue1396-pre-post-run-commands/dub.sdl (hunk 1 above)
Comment (andre2007) [reply test/issue1396-pre-post-run-commands/dub.sdl]: My impression was, the newline is needed for the grep command. But that is not true. I removed them.

File: test/issue1396-pre-post-run-commands.sh (hunk 2 above)
Comment (andre2007) [reply test/issue1396-pre-post-run-commands.sh]: Depending how the test suite parallel runner works this fails or not. I am in doubt to add here a more complex logic which might be needed or not.

File: test/issue1396-pre-post-run-commands/test.txt [hunk 3]
```diff
@@ -0,0 +1,4 @@
+pre-run
+
+post-run
+
```
[the remark this answers is not in the document]
Comment (andre2007) [reply test/issue1396-pre-post-run-commands/test.txt]: removed

File: source/dub/generators/build.d [hunk 4]
```diff
@@ -510,6 +510,11 @@ class BuildGenerator : ProjectGenerator {
 	private void runTarget(NativePath exe_file_path, in BuildSettings buildsettings, string[] run_args, GeneratorSettings settings)
 	{
 		if (buildsettings.targetType == TargetType.executable) {
+			if ( buildsettings.preRunCommands.length) {
```
[the remark this answers is not in the document]
Comment (andre2007) [reply source/dub/generators/build.d]: fixed

File: source/dub/generators/build.d [hunk 5]
```diff
@@ -538,6 +543,11 @@ class BuildGenerator : ProjectGenerator {
 				auto result = prg_pid.wait();
 				enforce(result == 0, "Program exited with code "~to!string(result));
 			}
+
+			if (buildsettings.postRunCommands.length) {
+				logInfo("Running post-run commands...");
+				runBuildCommands(buildsettings.postRunCommands, m_project.rootPackage, m_project, settings, buildsettings);
```
[the remark this answers is not in the document]
Comment (andre2007) [reply source/dub/generators/build.d]: At the moment, there is an existing  enforce call, therefore the post run commands are only called if return code is 0. 
Should post run commands be called even if return code is not equal 0?

[the remark this answers is not in the document]
Comment (andre2007) [reply source/dub/generators/build.d]: Environment variable DUB_TARGET_STATUS added, containing the status of the executable call.
Also the pre / post run commands are no longer executed for callback (dustmite).

Please have a look

File: changelog/pre-post-run-commands.dd [hunk 6]
```diff
@@ -0,0 +1,6 @@
+Pre/Post run commands added
+
+DUB now supports commands preRunCommands which are executed before the target run
+and postRunCommands which are executed after the target run.
+Environment variable DUB_TARGET_STATUS contains the target executable call status
```
[the remark this answers is not in the document]
Comment (andre2007) [reply changelog/pre-post-run-commands.dd]: fixed

File: source/dub/generators/build.d [hunk 7]
```diff
@@ -529,13 +529,23 @@ class BuildGenerator : ProjectGenerator {
...
+					runBuildCommands(buildsettings.preRunCommands, m_project.rootPackage, m_project, settings, buildsettings);
+				}
+				logInfo("Running %s %s", exe_path_string, run_args.join(" "));
 				auto prg_pid = spawnProcess(exe_path_string ~ run_args);
 				auto result = prg_pid.wait();
+				settings.targetStatus = result;
+				if (buildsettings.postRunCommands.length) {
+					logInfo("Running post-run commands...");
+					runBuildCommands(buildsettings.postRunCommands, m_project.rootPackage, m_project, settings, buildsettings);
+				}
```
[the remark this answers is not in the document]
Comment (andre2007) [reply source/dub/generators/build.d]: fixed

File: source/dub/generators/generator.d [hunk 8]
```diff
@@ -439,6 +439,7 @@ struct GeneratorSettings {
 	string buildType;
 	BuildSettings buildSettings;
 	BuildMode buildMode = BuildMode.separate;
+	int targetStatus;
```
[the remark this answers is not in the document]
Comment (andre2007) [reply source/dub/generators/generator.d]: fixed

File: source/dub/generators/generator.d [hunk 9]
```diff
@@ -439,6 +439,7 @@ struct GeneratorSettings {
 	string buildType;
 	BuildSettings buildSettings;
 	BuildMode buildMode = BuildMode.separate;
+	int targetExitStatus;
```
Comment (WebFreak001) [inline source/dub/generators/generator.d]: is GeneratorSettings really the proper place to put this? All the other environment variables are always set after all and targetExitStatus is only set once the program has run once, so it wouldn't make sense to include it in a pre-run call for example.

Also why is this called "target" exit status?

File: source/dub/generators/build.d [hunk 10]
```diff
@@ -529,19 +529,41 @@ class BuildGenerator : ProjectGenerator {
 				if (!exe_path_string.startsWith(".") && (exe_path_string.length < 2 || exe_path_string[1] != ':'))
 					exe_path_string = ".\\" ~ exe_path_string;
 			}
+			runPreRunCommands(m_project.rootPackage, m_project, settings, buildsettings);
 			logInfo("Running %s %s", exe_path_string, run_args.join(" "));
 			if (settings.runCallback) {
 				auto res = execute(exe_path_string ~ run_args);
 				settings.runCallback(res.status, res.output);
+				runPostRunCommands(m_project.rootPackage, m_project, settings, buildsettings);
```
Comment (WebFreak001) [inline source/dub/generators/build.d]: shouldn't targetExitStatus also be set here before the run callback?

Comment (andre2007) [reply source/dub/generators/build.d]: you are right, fixed

File: source/dub/generators/generator.d (hunk 9 above)
Comment (andre2007) [reply source/dub/generators/generator.d]: It seemed for me reasonable to name it "target" as e.g. the existing method also is called "runTarget". 
You are right, it is a compromise. I didn't wanted to add too much new complexity just to get the targetExitStatus as environment variable.

**Comment by WebFreak001** [comment]
oh yeah sorry, I think it's fine to merge now. I still don't like that the environment variable is in every run, but I think this makes the code the simplest and nobody will notice anyway.

**Pull request merged**